File: apps/actor-ios/ActorClient/CocoaTcpConnection.m
//
//  CocoaTcpConnection.m
//  ActorModel
//
//  Created by Антон Буков on 16.02.15.
//  Copyright (c) 2015 Anton Bukov. All rights reserved.
//

#import <zlib.h>
#import <CocoaAsyncSocket/GCDAsyncSocket.h>
#import "J2ObjC_source.h"
#import "im/actor/model/network/ConnectionEndpoint.h"
#import "im/actor/model/network/ConnectionCallback.h"
#import "im/actor/model/network/CreateConnectionCallback.h"
#import "AASerializationHelpers.h"
#import "CocoaTcpConnection.h"

@interface CocoaTcpConnection () <GCDAsyncSocketDelegate>

@property (nonatomic, strong) AMConnectionEndpoint *endpoint;
@property (nonatomic, strong) id<AMConnectionCallback> callback;
@property (nonatomic, strong) id<AMCreateConnectionCallback> createCallback;
@property (nonatomic, strong) GCDAsyncSocket *socket;

@property (nonatomic, strong) NSMutableData *buffer;
@property (nonatomic, strong) NSMutableData *outBuffer;
@property (nonatomic, assign) uint32_t packetIndex;

@end

@implementation CocoaTcpConnection

- (NSMutableData *)buffer
{
    if (_buffer == nil)
        _buffer = [NSMutableData data];
    return _buffer;
}

#pragma mark - GCD Async Socket

- (void)socket:(GCDAsyncSocket *)sock didConnectToHost:(NSString *)host port:(uint16_t)port
{
    NSLog(@"socket:didConnectToHost:port:");
    self.buffer = nil;
    if (self.endpoint.getType.ordinal == AMConnectionEndpoint_Type_TCP_TLS) {
        [sock startTLS:@{//(id)kCFStreamSSLAllowsExpiredCertificates:@NO,
                         //(id)kCFStreamSSLAllowsExpiredRoots:@NO,
                         //(id)kCFStreamSSLAllowsAnyRoot:@YES,
                         //(id)kCFStreamSSLValidatesCertificateChain:@YES,
                         (id)kCFStreamSSLPeerName:@"actor.im",
                         //(id)kCFStreamSSLLevel:(id)kCFStreamSocketSecurityLevelNegotiatedSSL,
                         }];
    } else {
        [self socketOk];
    }
}

- (void)socketDidSecure:(GCDAsyncSocket *)sock
{
    NSLog(@"socketDidSecure");
    [self socketOk];
}

- (void)socketOk
{
    [self.socket readDataWithTimeout:-1 tag:0];
    [self.createCallback onConnectionCreated:self];
}

- (void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag
{
    [self.buffer appendData:data];
    [self tryToGetMessageFromBuffer];
    [sock readDataWithTimeout:-1 tag:0];
}

- (void)socketDidDisconnect:(GCDAsyncSocket *)sock withError:(NSError *)err
{
    NSLog(@"socketDidDisconnect:withError: %@", err);
    [self.callback onConnectionDie];
}

#pragma mark - Tcp Connection

- (void)tryToGetMessageFromBuffer
{
    NSInteger offset = 0;
    const uint8_t *bytes = self.buffer.bytes;
    
    uint32_t packetLength;
    uint32_t packetIndex;
    if (self.buffer.length < sizeof(packetLength))
        return;
    packetLength = extract_uint(bytes, &offset);
    if (self.buffer.length < sizeof(packetLength) + packetLength)
        return;
    packetIndex = extract_uint(bytes, &offset);
    
    uint32_t checksum = (int32_t)crc32(0, bytes, (int32_t)(packetLength));
    uint32_t *checksumOriginalPtr = (uint32_t *)(bytes+4+packetLength-4);
    uint32_t checksumOriginal = ntohl(*checksumOriginalPtr);
    
    // If checksum valid
    if (checksum == checksumOriginal) {
        IOSByteArray *arr = [IOSByteArray arrayWithBytes:self.buffer.bytes count:self.buffer.length];
        [self.callback onMessage:arr withOffset:8 withLen:packetLength-8];
    } else {
        NSLog(@"CRC-32 error");
        [self close];
        return;
    }
    
    [self.buffer replaceBytesInRange:NSMakeRange(0, 4+packetLength) withBytes:NULL length:0];
    [self tryToGetMessageFromBuffer];
}

- (instancetype)initWithConnectionId:(jint)connectionId
                  connectionEndpoint:(AMConnectionEndpoint *)endpoint
                  connectionCallback:(id<AMConnectionCallback>)callback
                      createCallback:(id<AMCreateConnectionCallback>)createCallback
{
    if (self = [super init]) {
        self.endpoint = endpoint;
        self.callback = callback;
        self.createCallback = createCallback;
        self.socket = [[GCDAsyncSocket alloc] initWithDelegate:self delegateQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)];
        NSError *error;
        NSLog(@"connectToHost:onPort: %@ %@",self.endpoint.getHost,@(self.endpoint.getPort));
        [self.socket connectToHost:self.endpoint.getHost onPort:self.endpoint.getPort error:&error];
        if (error)
            NSLog(@"Socket connection error: %@", error);
    }
    return self;
}

- (void)post:(IOSByteArray *)data withOffset:(jint)offset withLen:(jint)len
{
    NSMutableData *buffer = [NSMutableData dataWithCapacity:4+4+len+4];
    
    insert_uint(buffer, 4 + (uint32_t)data.length + 4);
    insert_uint(buffer, self.packetIndex++);
    [buffer appendData:[data.toNSData subdataWithRange:NSMakeRange(offset, len)]];
    insert_uint(buffer, (uint32_t)crc32(0, buffer.bytes, (int32_t)buffer.length));
    
    if (!self.socket.isConnected) {
        return;
    }
    [self.socket writeData:buffer withTimeout:-1 tag:0];
}

- (jboolean)isClosed
{
    return self.socket.isDisconnected;
}

- (void)close
{
    [self.socket disconnect];
}

@end

File: apps/actor-ios/ActorClient/CocoaTcpConnection.swift
//
//  Copyright (c) 2015 Actor LLC. <https://actor.im>
//

import Foundation

class CocoaTcpConnectionFactory: NSObject, AMAsyncConnectionFactory {
    func createConnectionWithInt(connectionId: jint, withAMConnectionEndpoint endpoint: AMConnectionEndpoint!, withAMAsyncConnectionInterface connectionInterface: AMAsyncConnectionInterface!) -> AMAsyncConnection! {
        return CocoaTcpConnection(connectionId: Int(connectionId), AMConnectionEndpoint: endpoint, withAMAsyncConnectionInterface: connectionInterface)
    }
}

class CocoaTcpConnection: AMAsyncConnection, GCDAsyncSocketDelegate {

    let READ_HEADER = 1
    let READ_BODY = 2
    
    var TAG: String!
    var gcdSocket:GCDAsyncSocket? = nil
    
    var header: NSData?
    
    init(connectionId: Int, AMConnectionEndpoint endpoint: AMConnectionEndpoint!, withAMAsyncConnectionInterface connection: AMAsyncConnectionInterface!) {
        super.init(AMConnectionEndpoint: endpoint, withAMAsyncConnectionInterface: connection)
        TAG = "🎍ConnectionTcp#\(connectionId)"
    }
    
    override func doConnect() {
//        NSLog("\(TAG) connecting...")
        gcdSocket = GCDAsyncSocket(delegate: self, delegateQueue: dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0))
        var endpoint = getEndpoint()
        gcdSocket!.connectToHost(endpoint.getHost()!, onPort: UInt16(endpoint.getPort()), withTimeout: Double(AMManagedConnection_CONNECTION_TIMEOUT) / 1000.0, error: nil)
    }
    
    // Affer successful connection
    func socket(sock: GCDAsyncSocket!, didConnectToHost host: String!, port: UInt16) {
        if (UInt(self.getEndpoint().getType().ordinal()) == AMConnectionEndpoint_Type.TCP_TLS.rawValue) {
//            NSLog("\(TAG) Starring TLS Session...")
            sock.startTLS(nil)
        } else {
            startConnection()
        }
    }
    
    // After TLS successful
    func socketDidSecure(sock: GCDAsyncSocket!) {
//        NSLog("\(TAG) TLS Session started...")
        startConnection()
    }
    
    func startConnection() {
        gcdSocket?.readDataToLength(UInt(9), withTimeout: -1, tag: READ_HEADER)
        onConnected()
    }
    
    // On connection closed
    func socketDidDisconnect(sock: GCDAsyncSocket!, withError err: NSError!) {
//        NSLog("\(TAG) Connection closed...")
        onClosed()
    }
    
    func socket(sock: GCDAsyncSocket!, didReadData data: NSData!, withTag tag: Int) {
        if (tag == READ_HEADER) {
//            NSLog("\(TAG) Header received")
            self.header = data
            var size = data.readUInt32(5)
            gcdSocket?.readDataToLength(UInt(size + 4), withTimeout: -1, tag: READ_BODY)
        } else if (tag == READ_BODY) {
//            NSLog("\(TAG) Body received")
            var package = NSMutableData()
            package.appendData(self.header!)
            package.appendData(data)
            self.header = nil
            gcdSocket?.readDataToLength(UInt(9), withTimeout: -1, tag: READ_HEADER)
            
            onReceivedWithByteArray(package.toJavaBytes())
        } else {
            fatalError("Unknown tag in read data")
        }
    }
    
    override func doClose() {
        if (gcdSocket != nil) {
//            NSLog("\(TAG) Closing...")
            gcdSocket?.disconnect()
            gcdSocket = nil
        }
    }
    
    override func doSendWithByteArray(data: IOSByteArray!) {
//        NSLog("\(TAG) Sending package...")
        gcdSocket?.writeData(data.toNSData(), withTimeout: -1, tag: 0)
    }
}